File: Boj-Kotlin&Java/백준15486번퇴사2.java
package 백준;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 백준15486번퇴사2 {
    static StringTokenizer st;
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        int n=Integer.parseInt(br.readLine());
        int[] t=new int[n+1];
        int[] p=new int[n+1];
        int[] dp=new int[n+51];
        for(int i=0; i<n; i++){
            st=new StringTokenizer(br.readLine());
            t[i]=Integer.parseInt(st.nextToken());
            p[i]=Integer.parseInt(st.nextToken());
        }
        for(int i=0; i<n; i++){
            dp[i+t[i]]=Math.max(dp[i+t[i]],dp[i]+p[i]);
            dp[i+1]=Math.max(dp[i],dp[i+1]);
        }
        System.out.println(dp[n]);
    }
}

File: Kotlin/백준 15486번 퇴사2.kt
import kotlin.math.max

fun main(args: Array<String>) {
    val br=System.`in`.bufferedReader()
    val n=br.readLine().toInt()
    val t=Array(n){0}
    val p=Array(n){0}
    val dp=Array(n+50){0}
    for(i in 0 until n)
    {
        val input=br.readLine().split(" ").map { it.toInt() }
        t[i]=input[0]
        p[i]=input[1]
    }
    for(i in 0 until n)
    {
        dp[i+t[i]]= max(dp[i+t[i]],dp[i]+p[i])
        dp[i+1]= max(dp[i+1],dp[i])
    }
    println("${dp[n]}")
}